### ISSUE TITLE
Panic when trying to create an app that already exists

### ISSUE THREAD

**Issue by der-eismann** [argoproj/argo-cd#8184]
**Describe the bug**

We just upgraded from argocd 2.1.6 to 2.2.2 and now our deployment chatbot doesn't work anymore. What we usually do is:
- Call http://example.com/api/v1/applications/example to grab a "parent" application
- Set some metadata & specs like plugin
- Send the new manifest back via POST to http://example.com/api/v1/applications
- If there is an error and it contains `use upsert flag to force update` we try again with a PUT to http://example.com/api/v1/applications/example (workaround from #3464)

This used to work, but after the upgrade to argocd 2.2.2 we get the following response when sending the POST to the /api/v1/applications endpoint if the application already exists:
```
error: code 13: reflect.Value.Interface: cannot return value obtained from unexported field or method
```

argocd-server itself is printing the whole stack:
<details><summary>error stack</summary>
<p>

```python
error msg="Recovered from panic: reflect.Value.Interface: cannot return value obtained from unexported field or method
goroutine 4463 [running]:
runtime/debug.Stack(0xc00281d870, 0x29c2260, 0x3d08bc0)
	/usr/local/go/src/runtime/debug/stack.go:24 +0x9f
github.com/argoproj/argo-cd/v2/util/grpc.PanicLoggerUnaryServerInterceptor.func1.1(0xc000628690, 0xc00281f2c0)
	/go/src/github.com/argoproj/argo-cd/util/grpc/grpc.go:24 +0x76
panic(0x29c2260, 0x3d08bc0)
	/usr/local/go/src/runtime/panic.go:965 +0x1b9
reflect.valueInterface(0x29786a0, 0xc0003c4278, 0xa1, 0x2cb7b01, 0x2881e92, 0x10)
	/usr/local/go/src/reflect/value.go:1027 +0x169
reflect.Value.Interface(...)
	/usr/local/go/src/reflect/value.go:1016
github.com/r3labs/diff.(*Differ).diffBool(0xc000b89140, 0xc00190b540, 0x2, 0x2, 0x29786a0, 0xc0003c4278, 0xa1, 0x29786a0, 0xc0003c4368, 0xa1, ...)
	/go/pkg/mod/github.com/r3labs/diff@v1.1.0/diff_bool.go:25 +0x11f
github.com/r3labs/diff.(*Differ).diff(0xc000b89140, 0xc00190b540, 0x2, 0x2, 0x29786a0, 0xc0003c4278, 0xa1, 0x29786a0, 0xc0003c4368, 0xa1, ...)
	/go/pkg/mod/github.com/r3labs/diff@v1.1.0/diff.go:139 +0xd8b
github.com/r3labs/diff.(*Differ).diffStruct(0xc000b89140, 0xc000d44110, 0x1, 0x1, 0x2cb7b60, 0xc0003c4248, 0x99, 0x2cb7b60, 0xc0003c4338, 0x99, ...)
	/go/pkg/mod/github.com/r3labs/diff@v1.1.0/diff_struct.go:50 +0x4dd
github.com/r3labs/diff.(*Differ).diff(0xc000b89140, 0xc000d44110, 0x1, 0x1, 0x2cb7b60, 0xc0003c4248, 0x99, 0x2cb7b60, 0xc0003c4338, 0x99, ...)
	/go/pkg/mod/github.com/r3labs/diff@v1.1.0/diff.go:133 +0xf9b
github.com/r3labs/diff.(*Differ).diffStruct(0xc000b89140, 0x54c8618, 0x0, 0x0, 0x2d49d40, 0xc0003c41e0, 0x99, 0x2d49d40, 0xc0003c42d0, 0x99, ...)
	/go/pkg/mod/github.com/r3labs/diff@v1.1.0/diff_struct.go:50 +0x4dd
github.com/r3labs/diff.(*Differ).diff(0xc000b89140, 0x54c8618, 0x0, 0x0, 0x2d49d40, 0xc0003c41e0, 0x99, 0x2d49d40, 0xc0003c42d0, 0x99, ...)
	/go/pkg/mod/github.com/r3labs/diff@v1.1.0/diff.go:133 +0xf9b
github.com/r3labs/diff.Diff(0x2d49d40, 0xc0003c41e0, 0x2d49d40, 0xc0003c42d0, 0x7f7e726eb878, 0x50, 0xc00281e630, 0x4928bc, 0x52c8360)
	/go/pkg/mod/github.com/r3labs/diff@v1.1.0/diff.go:66 +0x18c
github.com/argoproj/argo-cd/v2/util/argo.GetDifferentPathsBetweenStructs(0x2d49d40, 0xc0003c41e0, 0x2d49d40, 0xc0003c42d0, 0x1, 0xc002014000, 0x47, 0xc0003c42d0, 0xc00178dd18)
	/go/src/github.com/argoproj/argo-cd/util/argo/argo.go:713 +0x67
github.com/argoproj/argo-cd/v2/util/argo.GenerateSpecIsDifferentErrorMessage(0x2ef6402, 0xb, 0x2d49d40, 0xc0003c41e0, 0x2d49d40, 0xc0003c42d0, 0x0, 0x0)
	/go/src/github.com/argoproj/argo-cd/util/argo/argo.go:704 +0xf1
github.com/argoproj/argo-cd/v2/server/application.(*Server).Create(0xc000864a90, 0x3da05b8, 0xc00242a480, 0xc00178d000, 0x0, 0x0, 0x0)
	/go/src/github.com/argoproj/argo-cd/server/application/application.go:216 +0x845
github.com/argoproj/argo-cd/v2/pkg/apiclient/application._ApplicationService_Create_Handler.func1(0x3da05b8, 0xc00242a480, 0x2d93240, 0xc00178d000, 0x52, 0x2f0c5a4, 0x14, 0x0)
	/go/src/github.com/argoproj/argo-cd/pkg/apiclient/application/application.pb.go:2813 +0x89
github.com/argoproj/argo-cd/v2/util/grpc.PanicLoggerUnaryServerInterceptor.func1(0x3da05b8, 0xc00242a480, 0x2d93240, 0xc00178d000, 0xc001b85680, 0xc00170f158, 0x0, 0x0, 0x0, 0x0)
	/go/src/github.com/argoproj/argo-cd/util/grpc/grpc.go:28 +0xa7
github.com/grpc-ecosystem/go-grpc-middleware.ChainUnaryServer.func1.1.1(0x3da05b8, 0xc00242a480, 0x2d93240, 0xc00178d000, 0x0, 0x0, 0x0, 0x0)
	/go/pkg/mod/github.com/grpc-ecosystem/go-grpc-middleware@v1.3.0/chain.go:25 +0x63
github.com/argoproj/argo-cd/v2/util/grpc.ErrorCodeGitUnaryServerInterceptor.func1(0x3da05b8, 0xc00242a480, 0x2d93240, 0xc00178d000, 0xc001b85680, 0xc001b85860, 0x0, 0x0, 0x0, 0x0)
	/go/src/github.com/argoproj/argo-cd/util/grpc/errors.go:83 +0x55
github.com/grpc-ecosystem/go-grpc-middleware.ChainUnaryServer.func1.1.1(0x3da05b8, 0xc00242a480, 0x2d93240, 0xc00178d000, 0x0, 0x29c2260, 0xc003241210, 0x0)
	/go/pkg/mod/github.com/grpc-ecosystem/go-grpc-middleware@v1.3.0/chain.go:25 +0x63
github.com/argoproj/argo-cd/v2/util/grpc.ErrorCodeK8sUnaryServerInterceptor.func1(0x3da05b8, 0xc00242a480, 0x2d93240, 0xc00178d000, 0xc001b85680, 0xc001b85880, 0x0, 0x0, 0x0, 0x0)
	/go/src/github.com/argoproj/argo-cd/util/grpc/errors.go:99 +0x55
github.com/grpc-ecosystem/go-grpc-middleware.ChainUnaryServer.func1.1.1(0x3da05b8, 0xc00242a480, 0x2d93240, 0xc00178d000, 0xc00178d000, 0x3da05b8, 0xc00242a480, 0x1)
	/go/pkg/mod/github.com/grpc-ecosystem/go-grpc-middleware@v1.3.0/chain.go:25 +0x63
github.com/argoproj/argo-cd/v2/util/grpc.PayloadUnaryServerInterceptor.func1(0x3da05b8, 0xc00242a480, 0x2d93240, 0xc00178d000, 0xc001b85680, 0xc001b858a0, 0xc000f860a0, 0xc000f86080, 0x3, 0xc000f07588)
	/go/src/github.com/argoproj/argo-cd/util/grpc/logging.go:80 +0x1e2
github.com/grpc-ecosystem/go-grpc-middleware.ChainUnaryServer.func1.1.1(0x3da05b8, 0xc00242a480, 0x2d93240, 0xc00178d000, 0x0, 0x40a71f, 0xc00003a000, 0x2b14640)
	/go/pkg/mod/github.com/grpc-ecosystem/go-grpc-middleware@v1.3.0/chain.go:25 +0x63
github.com/argoproj/argo-cd/v2/util/grpc.UserAgentUnaryServerInterceptor.func1(0x3da05b8, 0xc00242a480, 0x2d93240, 0xc00178d000, 0xc001b85680, 0xc001b858c0, 0x3da05b8, 0xc00242a480, 0x0, 0x0)
	/go/src/github.com/argoproj/argo-cd/util/grpc/useragent.go:22 +0xa5
github.com/grpc-ecosystem/go-grpc-middleware.ChainUnaryServer.func1.1.1(0x3da05b8, 0xc00242a480, 0x2d93240, 0xc00178d000, 0x0, 0x0, 0x7f7e4b8d36f0, 0xc0005e9620)
	/go/pkg/mod/github.com/grpc-ecosystem/go-grpc-middleware@v1.3.0/chain.go:25 +0x63
github.com/grpc-ecosystem/go-grpc-middleware/auth.UnaryServerInterceptor.func1(0x3da05b8, 0xc001a75ad0, 0x2d93240, 0xc00178d000, 0xc001b85680, 0xc001b858e0, 0x3, 0x3, 0x7f7e4b8d36f0, 0xc0005e9620)
	/go/pkg/mod/github.com/grpc-ecosystem/go-grpc-middleware@v1.3.0/auth/auth.go:47 +0x10e
github.com/grpc-ecosystem/go-grpc-middleware.ChainUnaryServer.func1.1.1(0x3da05b8, 0xc001a75ad0, 0x2d93240, 0xc00178d000, 0x26, 0xc000156960, 0x0, 0x0)
	/go/pkg/mod/github.com/grpc-ecosystem/go-grpc-middleware@v1.3.0/chain.go:25 +0x63
github.com/grpc-ecosystem/go-grpc-prometheus.(*ServerMetrics).UnaryServerInterceptor.func1(0x3da05b8, 0xc001a75ad0, 0x2d93240, 0xc00178d000, 0xc001b85680, 0xc001b85900, 0x0, 0x0, 0x0, 0x0)
	/go/pkg/mod/github.com/grpc-ecosystem/go-grpc-prometheus@v1.2.0/server_metrics.go:107 +0xb0
github.com/grpc-ecosystem/go-grpc-middleware.ChainUnaryServer.func1.1.1(0x3da05b8, 0xc001a75ad0, 0x2d93240, 0xc00178d000, 0x26, 0xc07088f875088f60, 0x16b715b24a3, 0x548a0c0)
	/go/pkg/mod/github.com/grpc-ecosystem/go-grpc-middleware@v1.3.0/chain.go:25 +0x63
github.com/grpc-ecosystem/go-grpc-middleware/logging/logrus.UnaryServerInterceptor.func1(0x3da05b8, 0xc001a75920, 0x2d93240, 0xc00178d000, 0xc001b85680, 0xc001b85920, 0x203000, 0x7f7e4b67c020, 0xc3, 0x2bc12e0)
	/go/pkg/mod/github.com/grpc-ecosystem/go-grpc-middleware@v1.3.0/logging/logrus/server_interceptors.go:31 +0x12f
github.com/grpc-ecosystem/go-grpc-middleware.ChainUnaryServer.func1.1.1(0x3da05b8, 0xc001a75920, 0x2d93240, 0xc00178d000, 0x20, 0x2c68420, 0x415600, 0x7f7e4b67c020)
	/go/pkg/mod/github.com/grpc-ecosystem/go-grpc-middleware@v1.3.0/chain.go:25 +0x63
github.com/argoproj/argo-cd/v2/server.bug21955WorkaroundInterceptor(0x3da05b8, 0xc001a75920, 0x2d93240, 0xc00178d000, 0xc001b85680, 0xc001b85940, 0x1fa73ba, 0x2c68420, 0xc001b85960, 0xc001b85680)
	/go/src/github.com/argoproj/argo-cd/server/server.go:1095 +0x177
github.com/grpc-ecosystem/go-grpc-middleware.ChainUnaryServer.func1.1.1(0x3da05b8, 0xc001a75920, 0x2d93240, 0xc00178d000, 0xc00007a800, 0x0, 0xc000f07a50, 0x40e3f8)
	/go/pkg/mod/github.com/grpc-ecosystem/go-grpc-middleware@v1.3.0/chain.go:25 +0x63
github.com/grpc-ecosystem/go-grpc-middleware.ChainUnaryServer.func1(0x3da05b8, 0xc001a75920, 0x2d93240, 0xc00178d000, 0xc001b85680, 0xc00170f158, 0xc001eccac0, 0x531be6, 0x2cced40, 0xc001a75920)
	/go/pkg/mod/github.com/grpc-ecosystem/go-grpc-middleware@v1.3.0/chain.go:34 +0xd7
github.com/argoproj/argo-cd/v2/pkg/apiclient/application._ApplicationService_Create_Handler(0x2e60600, 0xc000864a90, 0x3da05b8, 0xc001a75920, 0xc001170460, 0xc00061cde0, 0x3da05b8, 0xc001a75920, 0xc001907400, 0x3dc1110)
	/go/src/github.com/argoproj/argo-cd/pkg/apiclient/application/application.pb.go:2815 +0x150
google.golang.org/grpc.(*Server).processUnaryRPC(0xc0006d7b00, 0x3dc4878, 0xc000e9e600, 0xc000148f00, 0xc00061d2c0, 0x54576d0, 0x0, 0x0, 0x0)
	/go/pkg/mod/google.golang.org/grpc@v1.15.0/server.go:982 +0x4eb
google.golang.org/grpc.(*Server).handleStream(0xc0006d7b00, 0x3dc4878, 0xc000e9e600, 0xc000148f00, 0x0)
	/go/pkg/mod/google.golang.org/grpc@v1.15.0/server.go:1208 +0x122c
google.golang.org/grpc.(*Server).serveStreams.func1.1(0xc001248d80, 0xc0006d7b00, 0x3dc4878, 0xc000e9e600, 0xc000148f00)
	/go/pkg/mod/google.golang.org/grpc@v1.15.0/server.go:686 +0xab
created by google.golang.org/grpc.(*Server).serveStreams.func1
	/go/pkg/mod/google.golang.org/grpc@v1.15.0/server.go:684 +0xa5
```

</p>
</details>

What I also find interesting is that in the log entry before the stack dump there is also a panic:
```
level=info msg="received unary call /application.ApplicationService/Create" grpc.method=Create 
grpc.request.claims="{\"iat\":...,\"iss\":\"argocd\",\"jti\":\"...\",\"nbf\":...,\"sub\":\"...\"}" 
grpc.request.content="%!v(PANIC=String method: reflect.Value.Bytes of non-byte slice)" 
grpc.service=application.ApplicationService grpc.start_time="2022-01-14T17:01:21Z" span.kind=server system=grpc
```

**To Reproduce**

Try to create the same application twice by sending it to the http://example.com/api/v1/applications endpoint via POST:
```json
{
    "metadata": {
        "name": "prometheus-default",
        "namespace": "argocd",
        "annotations": {
            "LockState": "",
            "LockedBy": "",
            "kubectl.kubernetes.io/last-applied-configuration": "..."
        },
        "labels": {
            "app.kubernetes.io/component": "",
            "app.kubernetes.io/name": "prometheus",
            "app.kubernetes.io/version": "master",
            "example.com/release": "default"
        },
        "finalizers": [
            "resources-finalizer.argocd.argoproj.io"
        ]
    },
    "spec": {
        "source": {
            "repoURL": "git@github.com:example/k8s-prometheus.git",
            "path": "deployment/k8s/branch",
            "targetRevision": "master"
        },
        "destination": {
            "name": "in-cluster",
            "namespace": "default"
        },
        "project": "default",
        "syncPolicy": {
            "automated": {
                "prune": false,
                "selfHeal": false
            }
        }
    }
}
```

**Expected behavior**

As before:
```
existing application spec is different, use upsert flag to force update
```

**Version**

```shell
argocd: v2.2.2+03b17e0
  BuildDate: 2022-01-01T06:27:52Z
  GitCommit: 03b17e0233e64787ffb5fcf65c740cc2a20822ba
  GitTreeState: clean
  GoVersion: go1.16.11
  Compiler: gc
  Platform: linux/amd64
```

**Comment by jannfis** [comment]
Thanks for the detailed reproduction specs. I could reproduce it against v2.2.2 and HEAD.

Seems to be a regression introduced by https://github.com/argoproj/argo-cd/pull/7224

@pasha-codefresh Can you have a look please?

**Comment by pasha-codefresh** [comment]
Sure @jannfis , im taking it

**Comment by jannfis** [comment]
The fix for this has been cherry-picked into `release-2.2` branch and should be released with v2.2.3.

**Comment by der-eismann** [comment]
Awesome, thanks for the quick fix!
